charms/reactive/altrelations.py

```diff
@@ -41,9 +49,574 @@ def from_name(cls, relation_name):
+    """
+    A :class:`~charms.reactive.altrelations.KeyList` view of
+    :class:`~charms.reactive.altrelations.RelatedUnit` items, with properties to
+    access a merged view of all of the units' data.
+
+    You can iterate over this view like any other list, or you can look up units
+    by their ``unit_name``.  Units will be in order by relation ID and unit
+    name.  If a given unit name occurs more than once, accessing it by
+    ``unit_name`` will return the one from the lowest relation ID::
+
+        # given the following relations...
+        {
+            'endpoint:1': {
+                'unit/1': {
+                    'key0': 'value0_1_1',
+                    'key1': 'value1_1_1',
+                },
+                'unit/0': {
+                    'key0': 'value0_1_0',
+                    'key1': 'value1_1_0',
+                },
+            },
+            'endpoint:0': {
+                'unit/1': {
+                    'key0': 'value0_0_1',
+                    'key2': 'value2_0_1',
+                },
+            },
+        }
+
+        from_all = endpoint.all_units['unit/1']
+        by_rel = endpoint.relations['endpoint:0'].units['unit/1']
+        by_index = endpoint.relations[0].units[1]
+        assert from_all is by_rel
+        assert by_rel is by_index
+
+    You can also use the
+    :attr:`~charms.reactive.altrelations.CombinedUnitsView.received` or
+    :attr:`~charms.reactive.altrelations.CombinedUnitsView.received_json`
+    properties just like you would on a single unit.  The data in these
+    collections will have all of the data from every unit, with units with the
+    lowest relation ID and unit name taking precedence if multiple units have
+    set a given field.  For example::
+
+        # given the same relations as above...
+
+        # the values across all relations would be:
+        assert endpoint.all_units.received['key0'] == 'value0_0_0'
+        assert endpoint.all_units.received['key1'] == 'value1_1_0'
+        assert endpoint.all_units.received['key2'] == 'value2_0_1'
+
+        # across individual relations:
+        assert endpoint.relations[0].units.received['key0'] == 'value0_0_1'
+        assert endpoint.relations[0].units.received['key1'] == None
+        assert endpoint.relations[0].units.received['key2'] == 'value2_0_1'
+        assert endpoint.relations[1].units.received['key0'] == 'value0_1_0'
+        assert endpoint.relations[1].units.received['key1'] == 'value1_1_0'
+        assert endpoint.relations[1].units.received['key2'] == None
+
+        # and of course you an access them by individual unit
+        assert endpoint.relations['endpoint:1'].units['unit/1'].received['key0'] \
+                == 'value0_1_1'
+
+    """
+    def __init__(self, items):
+        super().__init__(sorted(items, key=lambda i: (i.relation.relation_id,
+                                                      i.unit_name)),
+                         key='unit_name')
```

Comment:
This is sorting by relation id as a string, which will put db:10 before db:2. Similarly the unit_name, which you claim later is order by unit number. Consider key=lambda i: (int(':'.split(i.relation.relation_id)[1]), int('/'.split(i.unit_name)[1]))

Reply:
Good catch, thanks.